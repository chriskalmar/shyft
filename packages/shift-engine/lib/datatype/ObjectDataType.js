
import {
  passOrThrow,
  resolveFunctionMap,
  isMap,
  isFunction,
} from '../util';

import { isEntity } from '../entity/Entity';
import { isDataType } from './DataType';
import ComplexDataType, { isComplexDataType } from './ComplexDataType';


class ObjectDataType extends ComplexDataType {

  constructor (setup = {}) {

    super()

    const {
      name,
      attributes,
    } = setup

    passOrThrow(name, () => 'Missing object data type name')
    passOrThrow(attributes, () => `Missing attributes for object data type '${name}'`)

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () => `Object data type '${name}' needs an attribute definition as a map or a function returning such a map`
    )

    this.name = name
    this._attributesMap = attributes
  }


  getAttributes () {
    if (this._attributes) {
      return this._attributes
    }

    const ret = this._attributes = this._processAttributeMap()
    return ret
  }


  _processAttribute (rawAttribute, attributeName) {

    if (rawAttribute instanceof ObjectDataType) {
      rawAttribute.getAttributes()
      return rawAttribute
    }

    const attribute = {
      ...rawAttribute,
      required: !!rawAttribute.required,
      name: attributeName
    }

    passOrThrow(attribute.description, () => `Missing description for '${this.name}.${attributeName}'`)

    passOrThrow(
      isDataType(attribute.type) || isEntity(attribute.type),
      () => `'${this.name}.${attributeName}' has invalid data type '${String(attribute.type)}'`
    )

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`
    )

    passOrThrow(
      !attribute.defaultValue || isFunction(attribute.defaultValue),
      () => `'${this.name}.${attributeName}' has an invalid defaultValue function'`
    )

    return attribute
  }


  _processAttributeMap () {

    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () => `Attribute definition function for object data type '${this.name}' does not return a map`
    )


    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Object data type '${this.name}' has no attributes defined`
    )

    const resultAttributes = {}

    attributeNames.forEach((attributeName) => {
      resultAttributes[ attributeName ] = this._processAttribute(attributeMap[ attributeName ], attributeName)
    })

    return resultAttributes
  }


  toString() {
    return this.name
  }

}


export default ObjectDataType


export const isObjectDataType = (obj) => {
  return (obj instanceof ObjectDataType)
}

