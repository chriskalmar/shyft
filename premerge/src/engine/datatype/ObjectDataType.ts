import { passOrThrow, resolveFunctionMap, isMap, isFunction } from '../util';
import { isEntity } from '../entity/Entity';
import { isDataType } from './DataType';
import { ComplexDataType, isComplexDataType } from './ComplexDataType';
import {
  AttributesMap,
  AttributesMapGenerator,
  AttributeBase,
} from '../attribute/Attribute';

export type ObjectDataTypeSetupType = {
  name: string;
  description: string;
  attributes: AttributesMap | AttributesMapGenerator;
};

export class ObjectDataType extends ComplexDataType {
  name: string;
  description: string;
  attributes: AttributesMap;
  _attributesMap: AttributesMap | AttributesMapGenerator;
  _attributes: AttributesMap;

  constructor(setup: ObjectDataTypeSetupType = {} as ObjectDataTypeSetupType) {
    // constructor(setup: ObjectDataTypeSetupType = <ObjectDataTypeSetupType>{}) {
    super();

    const { name, description, attributes } = setup;

    passOrThrow(name, () => 'Missing object data type name');
    passOrThrow(
      description,
      () => `Missing description for object data type '${name}'`,
    );
    passOrThrow(
      attributes,
      () => `Missing attributes for object data type '${name}'`,
    );

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () =>
        `Object data type '${name}' needs an attribute definition as a map or a function returning such a map`,
    );

    this.name = name;
    this.description = description;
    this._attributesMap = attributes;
  }

  getAttributes(): AttributesMap {
    if (this._attributes) {
      return this._attributes;
    }

    const ret = (this._attributes = this._processAttributeMap());
    return ret;
  }

  _processAttribute(
    rawAttribute: AttributeBase,
    attributeName: string,
  ): AttributeBase {
    if (isFunction(rawAttribute.type)) {
      const rawAttributeTypeFn = rawAttribute.type as Function;
      rawAttribute.type = rawAttributeTypeFn({
        name: attributeName,
        description: rawAttribute.description,
      });
    }

    const attribute = {
      ...rawAttribute,
      required: !!rawAttribute.required,
      name: attributeName,
    };

    passOrThrow(
      attribute.description,
      () => `Missing description for '${this.name}.${attributeName}'`,
    );

    passOrThrow(
      isDataType(attribute.type) ||
        isComplexDataType(attribute.type) ||
        isEntity(attribute.type),
      () =>
        `'${this.name}.${attributeName}' has invalid data type '${String(
          attribute.type,
        )}'`,
    );

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`,
    );

    if (attribute.defaultValue) {
      // enforce mandatory param if defaultValue provided
      attribute.required = true;

      passOrThrow(
        isFunction(attribute.defaultValue),
        () =>
          `'${this.name}.${attributeName}' has an invalid defaultValue function'`,
      );

      passOrThrow(
        !isComplexDataType(attribute.type),
        () =>
          `Complex data type '${this.name}.${attributeName}' cannot have a defaultValue function'`,
      );
    }

    passOrThrow(
      !attribute.validate || isFunction(attribute.validate),
      () => `'${this.name}.${attributeName}' has an invalid validate function'`,
    );

    return attribute;
  }

  _processAttributeMap(): AttributesMap {
    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () =>
        `Attribute definition function for object data type '${this.name}' does not return a map`,
    );

    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Object data type '${this.name}' has no attributes defined`,
    );

    const resultAttributes = {};

    attributeNames.forEach((attributeName) => {
      resultAttributes[attributeName] = this._processAttribute(
        attributeMap[attributeName],
        attributeName,
      );
    });

    return resultAttributes;
  }

  validate = (value: any): void => {
    if (value) {
      passOrThrow(
        isMap(value),
        () => `Object data type '${this.name}' expects an object`,
      );
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isObjectDataType = (obj: any): boolean => {
  return obj instanceof ObjectDataType;
};

export const buildObjectDataType = (obj: {
  // attributes: AttributesMap | AttributesMapGenerator | AttributeBase;
  attributes: any;
}): Function => {
  return ({ name, description }): ObjectDataType =>
    new ObjectDataType({
      description,
      ...obj,
      name,
    });
};
