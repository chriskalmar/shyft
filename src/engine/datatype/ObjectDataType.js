import { passOrThrow, resolveFunctionMap, isMap, isFunction } from '../util';
import { isEntity } from '../entity/Entity';
import { isDataType } from './DataType';
import { ComplexDataType, isComplexDataType } from './ComplexDataType';

export class ObjectDataType extends ComplexDataType {
  constructor(setup = {}) {
    super();

    const { name, description, attributes } = setup;

    passOrThrow(name, () => 'Missing object data type name');
    passOrThrow(
      description,
      () => `Missing description for object data type '${name}'`,
    );
    passOrThrow(
      attributes,
      () => `Missing attributes for object data type '${name}'`,
    );

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () =>
        `Object data type '${name}' needs an attribute definition as a map or a function returning such a map`,
    );

    this.name = name;
    this.description = description;
    this._attributesMap = attributes;
  }

  getAttributes() {
    if (this._attributes) {
      return this._attributes;
    }

    const ret = (this._attributes = this._processAttributeMap());
    return ret;
  }

  _processAttribute(rawAttribute, attributeName) {
    if (isFunction(rawAttribute.type)) {
      rawAttribute.type = rawAttribute.type({
        name: attributeName,
        description: rawAttribute.description,
      });
    }

    const attribute = {
      ...rawAttribute,
      required: !!rawAttribute.required,
      name: attributeName,
    };

    passOrThrow(
      attribute.description,
      () => `Missing description for '${this.name}.${attributeName}'`,
    );

    passOrThrow(
      isDataType(attribute.type) ||
        isComplexDataType(attribute.type) ||
        isEntity(attribute.type),
      () =>
        `'${this.name}.${attributeName}' has invalid data type '${String(
          attribute.type,
        )}'`,
    );

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`,
    );

    if (attribute.defaultValue) {
      // enforce mandatory param if defaultValue provided
      attribute.required = true;

      passOrThrow(
        isFunction(attribute.defaultValue),
        () =>
          `'${
            this.name
          }.${attributeName}' has an invalid defaultValue function'`,
      );

      passOrThrow(
        !isComplexDataType(attribute.type),
        () =>
          `Complex data type '${
            this.name
          }.${attributeName}' cannot have a defaultValue function'`,
      );
    }

    passOrThrow(
      !attribute.validate || isFunction(attribute.validate),
      () => `'${this.name}.${attributeName}' has an invalid validate function'`,
    );

    return attribute;
  }

  _processAttributeMap() {
    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () =>
        `Attribute definition function for object data type '${
          this.name
        }' does not return a map`,
    );

    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Object data type '${this.name}' has no attributes defined`,
    );

    const resultAttributes = {};

    attributeNames.forEach(attributeName => {
      resultAttributes[attributeName] = this._processAttribute(
        attributeMap[attributeName],
        attributeName,
      );
    });

    return resultAttributes;
  }

  validate = value => {
    if (value) {
      passOrThrow(
        isMap(value),
        () => `Object data type '${this.name}' expects an object`,
      );
    }
  };

  toString() {
    return this.name;
  }
}

export const isObjectDataType = obj => {
  return obj instanceof ObjectDataType;
};

export const buildObjectDataType = obj => {
  return ({ name, description }) =>
    new ObjectDataType({
      description,
      ...obj,
      name,
    });
};
