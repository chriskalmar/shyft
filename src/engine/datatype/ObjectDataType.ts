import { passOrThrow, resolveFunctionMap, isMap, isFunction } from '../util';
import { isEntity } from '../entity/Entity';
import { DataTypeFunction, DataTypeValidateType, isDataType } from './DataType';
import { ComplexDataType, isComplexDataType } from './ComplexDataType';
import {
  AttributesMap,
  AttributesMapGenerator,
  AttributeBase,
  AttributesSetupMap,
} from '../attribute/Attribute';

export type ObjectDataTypeSetupType = {
  name: string;
  description: string;
  attributes: AttributesSetupMap | AttributesMapGenerator;
};

export class ObjectDataType extends ComplexDataType {
  name: string;
  description: string;
  attributes: AttributesMap;
  private _attributesMap: AttributesSetupMap | AttributesMapGenerator;
  private _attributes: AttributesMap;

  constructor(setup: ObjectDataTypeSetupType = {} as ObjectDataTypeSetupType) {
    // constructor(setup: ObjectDataTypeSetupType = <ObjectDataTypeSetupType>{}) {
    super();

    const { name, description, attributes } = setup;

    passOrThrow(name, () => 'Missing object data type name');
    passOrThrow(
      description,
      () => `Missing description for object data type '${name}'`,
    );
    passOrThrow(
      attributes,
      () => `Missing attributes for object data type '${name}'`,
    );

    passOrThrow(
      isMap(attributes) || isFunction(attributes),
      () =>
        `Object data type '${name}' needs an attribute definition as a map or a function returning such a map`,
    );

    this.name = name;
    this.description = description;
    this._attributesMap = attributes;
  }

  getAttributes(): AttributesMap {
    if (this._attributes) {
      return this._attributes;
    }

    const ret = (this._attributes = this._processAttributeMap());
    return ret;
  }

  _processAttribute(
    rawAttribute: AttributeBase,
    attributeName: string,
  ): AttributeBase {
    if (isFunction(rawAttribute.type)) {
      const rawAttributeTypeFn = rawAttribute.type;
      rawAttribute.type = rawAttributeTypeFn({
        setup: {
          name: attributeName,
          description: rawAttribute.description,
        },
      });
    }

    const attribute = {
      ...rawAttribute,
      required: !!rawAttribute.required,
      name: attributeName,
    };

    passOrThrow(
      attribute.description,
      () => `Missing description for '${this.name}.${attributeName}'`,
    );

    passOrThrow(
      isDataType(attribute.type) ||
        isComplexDataType(attribute.type) ||
        isEntity(attribute.type),
      () =>
        `'${this.name}.${attributeName}' has invalid data type '${String(
          attribute.type,
        )}'`,
    );

    passOrThrow(
      !attribute.resolve || isFunction(attribute.resolve),
      () => `'${this.name}.${attributeName}' has an invalid resolve function'`,
    );

    if (attribute.defaultValue) {
      // enforce mandatory param if defaultValue provided
      attribute.required = true;

      passOrThrow(
        isFunction(attribute.defaultValue),
        () =>
          `'${this.name}.${attributeName}' has an invalid defaultValue function'`,
      );

      passOrThrow(
        !isComplexDataType(attribute.type),
        () =>
          `Complex data type '${this.name}.${attributeName}' cannot have a defaultValue function'`,
      );
    }

    passOrThrow(
      !attribute.validate || isFunction(attribute.validate),
      () => `'${this.name}.${attributeName}' has an invalid validate function'`,
    );

    return attribute;
  }

  _processAttributeMap(): AttributesMap {
    // if it's a function, resolve it to get that map
    const attributeMap = resolveFunctionMap(this._attributesMap);

    passOrThrow(
      isMap(attributeMap),
      () =>
        `Attribute definition function for object data type '${this.name}' does not return a map`,
    );

    const attributeNames = Object.keys(attributeMap);
    passOrThrow(
      attributeNames.length > 0,
      () => `Object data type '${this.name}' has no attributes defined`,
    );

    const resultAttributes = {};

    attributeNames.forEach((attributeName) => {
      resultAttributes[attributeName] = this._processAttribute(
        attributeMap[attributeName],
        attributeName,
      );
    });

    return resultAttributes;
  }

  validate: DataTypeValidateType = ({ value }) => {
    if (value) {
      if (!isMap(value)) {
        throw new Error(`Object data type '${this.name}' expects an object`);
      }
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isObjectDataType = (obj: unknown): obj is ObjectDataType => {
  return obj instanceof ObjectDataType;
};

export const buildObjectDataType = (obj: {
  attributes: AttributesSetupMap | AttributesMapGenerator;
}): DataTypeFunction => {
  return ({ setup: { name, description } }): ObjectDataType =>
    new ObjectDataType({
      description,
      attributes: obj.attributes,
      name,
    });
};
