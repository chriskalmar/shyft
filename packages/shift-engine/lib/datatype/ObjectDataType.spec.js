
import ObjectDataType, { isObjectDataType } from './ObjectDataType';

import {
  passOrThrow,
} from '../util';

import {
  DataTypeID,
  DataTypeString,
} from './dataTypes';


describe('ObjectDataType', () => {

  it('should have a name', () => {

    function fn() {
      new ObjectDataType() // eslint-disable-line no-new
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a description', () => {

    function fn() {
      new ObjectDataType({ // eslint-disable-line no-new
        name: 'example'
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should have a map of attributes', () => {

    function fn() {
      new ObjectDataType({ // eslint-disable-line no-new
        name: 'Example',
        description: 'Just some description',
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should return it\'s name', () => {

    const objectDataType = new ObjectDataType({
      name: 'someObjectDataTypeName',
      description: 'Just some description',
      attributes: {},
    })

    expect(objectDataType.name).toBe('someObjectDataTypeName');
    expect(String(objectDataType)).toBe('someObjectDataTypeName');

  })



  it('should accept only maps or functions as attributes definition', () => {

    function fn() {
      new ObjectDataType({ // eslint-disable-line no-new
        name: 'Example',
        description: 'Just some description',
        attributes: [ 2, 7, 13 ]
      })
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should reject non-map results of attribute definition functions', () => {

    function fn() {
      const objectDataType = new ObjectDataType({
        name: 'Example',
        description: 'Just some description',
        attributes: () => {
          return [ 2, 7, 13 ]
        }
      })

      objectDataType.getAttributes()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should reject empty attribute maps', () => {

    function fn() {
      const objectDataType = new ObjectDataType({
        name: 'Example',
        description: 'Just some description',
        attributes: {}
      })

      objectDataType.getAttributes()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should reject attributes with missing descriptions', () => {

    function fn() {
      const objectDataType = new ObjectDataType({
        name: 'Example',
        description: 'Just some description',
        attributes: {
          name: {
            type: DataTypeString,
          },
        }
      })

      objectDataType.getAttributes()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should reject attributes with invalid data types', () => {

    function fn() {
      const objectDataType = new ObjectDataType({
        name: 'Example',
        description: 'Just some description',
        attributes: {
          name: {
            type: {},
            description: 'Just some description',
          },
        }
      })

      objectDataType.getAttributes()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should reject attributes with invalid resolve functions', () => {

    function fn() {
      const objectDataType = new ObjectDataType({
        name: 'Example',
        description: 'Just some description',
        attributes: {
          name: {
            type: DataTypeString,
            description: 'Just some description',
            resolve: 123456,
          },
        }
      })

      objectDataType.getAttributes()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should reject attributes with invalid defaultValue functions', () => {

    function fn() {
      const objectDataType = new ObjectDataType({
        name: 'Example',
        description: 'Just some description',
        attributes: {
          name: {
            type: DataTypeString,
            description: 'Just some description',
            defaultValue: 123456,
          },
        }
      })

      objectDataType.getAttributes()
    }

    expect(fn).toThrowErrorMatchingSnapshot();

  })


  it('should cache the attributes map after initial processing', () => {

    const objectDataType = new ObjectDataType({
      name: 'SomeName',
      description: 'Just some description',
      attributes: {
        id: {
          type: DataTypeID,
          description: 'ID of item',
        },
      }
    })

    const attributes = objectDataType.getAttributes()
    const attributesAgain = objectDataType.getAttributes()

    expect(attributes).toEqual(attributesAgain)

  })


  it('should handle nested object data types', () => {

    const objectDataType = new ObjectDataType({
      name: 'SomeName',
      description: 'Just some description',
      attributes: {
        id: {
          type: DataTypeID,
          description: 'ID of item',
        },
        name: {
          type: DataTypeString,
          description: 'Just another description'
        },
        nested: {
          type: new ObjectDataType({
            name: 'SomeNestedName',
            description: 'Just some description',
            attributes: {
              randomInput: {
                type: DataTypeString,
                description: 'One more description'
              },
            }
          })
        }
      }
    })

    const attributes = objectDataType.getAttributes()
    const attributesAgain = objectDataType.getAttributes()

    expect(attributes).toEqual(attributesAgain)

    expect(attributes.id.type).toBe(DataTypeID);
    expect(attributes.name.type).toBe(DataTypeString);

    const attributesNested = attributes.nested.type.getAttributes()

    expect(attributesNested.randomInput.type).toBe(DataTypeString);

  })


  describe('should have a list of attributes', () => {

    it('as a map', () => {

      const objectDataType = new ObjectDataType({
        name: 'SomeName',
        description: 'Just some description',
        attributes: {
          id: {
            type: DataTypeID,
            description: 'ID of item',
          },
          name: {
            type: DataTypeString,
            description: 'Just another description'
          },
        }
      })

      const attributes = objectDataType.getAttributes()

      expect(attributes.id.type).toBe(DataTypeID);
      expect(attributes.name.type).toBe(DataTypeString);

    })


    it('as a function returning a map', () => {

      const objectDataType = new ObjectDataType({
        name: 'SomeName',
        description: 'Just some description',
        attributes: () => {
          return {
            id: {
              type: DataTypeID,
              description: 'ID of item',
            },
            name: {
              type: DataTypeString,
              description: 'Just another description'
            },
          }
        }
      })

      const attributes = objectDataType.getAttributes()

      expect(attributes.id.type).toBe(DataTypeID);
      expect(attributes.name.type).toBe(DataTypeString);

    })

  })



  describe('isObjectDataType', () => {


    it('should recognize objects of type ObjectDataType', () => {

      const objectDataType = new ObjectDataType({
        name: 'someObjectDataTypeName',
        description: 'Just some description',
        attributes: {},
      })

      function fn() {
        passOrThrow(
          isObjectDataType(objectDataType),
          () => 'This error will never happen'
        )
      }

      expect(fn).not.toThrow()

    })


    it('should recognize non-DataType objects', () => {

      function fn() {
        passOrThrow(
          isObjectDataType({}) ||
          isObjectDataType(function test() {}) ||
          isObjectDataType(Error),
          () => 'Not a DataType object'
        )
      }


      expect(fn).toThrowErrorMatchingSnapshot();

    })

  })


})
