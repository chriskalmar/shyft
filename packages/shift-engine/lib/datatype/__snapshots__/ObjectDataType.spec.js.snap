// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectDataType isObjectDataType should recognize non-DataType objects 1`] = `"Not a DataType object"`;

exports[`ObjectDataType should accept only maps or functions as attributes definition 1`] = `"Object data type 'Example' needs an attribute definition as a map or a function returning such a map"`;

exports[`ObjectDataType should have a description 1`] = `"Missing description for object data type 'example'"`;

exports[`ObjectDataType should have a map of attributes 1`] = `"Missing attributes for object data type 'Example'"`;

exports[`ObjectDataType should have a name 1`] = `"Missing object data type name"`;

exports[`ObjectDataType should reject attributes with invalid data types 1`] = `"'Example.name' has invalid data type '[object Object]'"`;

exports[`ObjectDataType should reject attributes with invalid defaultValue functions 1`] = `"'Example.name' has an invalid defaultValue function'"`;

exports[`ObjectDataType should reject attributes with invalid resolve functions 1`] = `"'Example.name' has an invalid resolve function'"`;

exports[`ObjectDataType should reject attributes with missing descriptions 1`] = `"Missing description for 'Example.name'"`;

exports[`ObjectDataType should reject empty attribute maps 1`] = `"Object data type 'Example' has no attributes defined"`;

exports[`ObjectDataType should reject non-map results of attribute definition functions 1`] = `"Attribute definition function for object data type 'Example' does not return a map"`;
